src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs

```diff
@@ -181,10 +199,22 @@ public void Connect(string hostname, int port)
                             Debug.Assert(address.AddressFamily == AddressFamily.InterNetwork || address.AddressFamily == AddressFamily.InterNetworkV6);
                             if ((address.AddressFamily == AddressFamily.InterNetwork && Socket.OSSupportsIPv4) || Socket.OSSupportsIPv6)
                             {
-                                tmpSocket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
-                                tmpSocket.Connect(address, port);
-                                _clientSocket = tmpSocket;
-                                tmpSocket = null;
+                                var socket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                                Volatile.Write(ref _clientSocket, socket);
+                                if (Volatile.Read(ref _cleanedUp) == 1)
```

Comment:
Can you elaborate on why we have all of these volatile reads and writes here?

Reply:
This is dealing with a race condition between Connect and Dispose.
When Dispose is called while we have created a new Socket here, we ensure that Socket is Disposed and the Connect throws ObjectDisposedException.